### ISSUE TITLE
BatchNorm generates NaN moving_variance on GPU with fused set to True for some inputs

### ISSUE THREAD

**Issue by jnd77** [tensorflow/tensorflow#34062]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 1.15.0
- Python version: 3.6
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory: the one on Colab

**Describe the current behavior**

I have an input of shape (1, 1, 1, num_channels), and I run it through a tf.keras.layers.BatchNormalization in training mode.

When running on CPU (fused or not) or GPU (not fused), the batch norm has the expected moving_variance of [0.99, 0.99 ...]

When running on GPU with fused=True, the moving_variance is [nan, nan, ...]

**Describe the expected behavior**

The moving variance should not be nan

**Code to reproduce the issue**
Check this [gist](https://colab.research.google.com/gist/jnd77/38ae530b17ee84e7b3907e3010c8529a/untitled0.ipynb)

**Other info / logs**
It works fine on CPU.

**Comment by dubesar** [comment]
Hi !!
The common reason for NaN can be the following :

1. Gradient Blow up
2. Bad learning rate policy and params
3. Faulty Loss function
4. stride larger than kernel size in "Pooling" layer

If possible check for the above problems.

**Comment by jnd77** [comment]
Hi. Thanks for the advice.
As you can see in the gist, I'm not running any model or training. It happens only for certain shapes of inputs (width and height equal to 1, but maybe for some others), so I guess there is a bug in the fused op on GPU for this corner case ...

**Comment by ravikyram** [comment]
I have tried on colab with TF version 1.15 on GPU with  `fused=True`,`fused=False`  and was able to reproduce the issue.Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/40cbe108f064e2d38bbda7b215f4e69a/untitled0.ipynb).However i am not seeing the issue with CPU. Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/ae83c64e7650839fbfb1c8e86c7eda2c/untitled343.ipynb).Thanks!

**Comment by karmel** [comment]
@jnd77 -- a lot has changed since this issue was first posted. Is this still an issue in TF v2? Can you try with tf-nightly or the most recent release of tensorflow?

**Comment by jnd77** [comment]
@karmel --  still here. Updated the [gist](https://colab.research.google.com/gist/jnd77/9a9a60621f9382b99e2dcb63eeae8af5/untitled0.ipynb) with TF 2.1.

**Comment by reedwm** [comment]
It appears `fused=False` uses the biased variance (meaning, [Bessel's correction](https://en.wikipedia.org/wiki/Bessel%27s_correction) is not used) while `fused=True` uses the unbiased variance. The fact there is a difference is bad, but that is a separate issue. If you take the unbiased variance of a single element, you compute 0/0, so arguably the correct result is getting NaNs. This is why passing `fused=True` gets NaNs.

The CPU has a special check if there is only one batch element [here](https://github.com/tensorflow/tensorflow/blob/9cc8c98a6017f7337d6b61f53118bcaa222bcd37/tensorflow/core/kernels/fused_batch_norm_op.cc#L176) which is why this doesn't occur on the CPU.

**Comment by ben-davidson-6** [comment]
This is also still an issue in tf 2.2.

``` python
import numpy as np
import tensorflow as tf

import os

x = np.random.randn(2, 1, 1, 1).astype(np.float32)
c = tf.constant(x)
c.set_shape([None, None, None, None])

bn = tf.keras.layers.BatchNormalization()
bn(x, training=True)
print(bn.moving_variance) # nan

bn = tf.keras.layers.BatchNormalization(epsilon=100)
bn(x, training=True)
print(bn.moving_variance) # nan

bn = tf.keras.layers.BatchNormalization(fused=False)
bn(x, training=True)
print(bn.moving_variance) # good

with tf.device('/cpu:0'):
    bn = tf.keras.layers.BatchNormalization()
    bn(x, training=True)
    print(bn.moving_variance)  # good
```

**Comment by Saduf2019** [comment]
@ben-davidson-6 
I ran the code shared on tf-nightly, please refer to this [gist here](https://colab.research.google.com/gist/Saduf2019/6f22e28af9bdf9b3a3044322dfcd7e05/untitled431.ipynb) and let us know if the issue persist.

**Comment by jnd77** [comment]
@Saduf2019 Thanks. It works for me.

**Comment by Saduf2019** [comment]
@jnd77 
Thank you for the update, glad its resolved. Please move the issue to closed status.

**Comment by zetez** [comment]
@jnd77 @Saduf2019 
The issue has not been resolved.
The reason it appears to be resolved in tf-nightly is because it is running on the CPU only.
I installed tf-nightly-gpu(2.5.0-dev20201029) on a local machine and the issue still persists.

**Comment by Saduf2019** [comment]
> @jnd77 @Saduf2019
> The issue has not been resolved.
> The reason it appears to be resolved in tf-nightly is because it is running on the CPU only.
> I installed tf-nightly-gpu(2.5.0-dev20201029) on a local machine and the issue still persists.

Please refer to the gist, the run time used is GPU. Please create a new issue if you face any using a new template.

**Comment by beelzmon** [comment]
Hello all,

I had the exact same issue, I had to change my network and put in a Flatten layer before the BN, and then a Reshape() as needed after it as a work around. This seems to have helped me, i don't know if it will help others. Hasn't affected effectiveness badly for me.
